### ISSUE TITLE
Logging based on user

### ISSUE THREAD

**Issue by aharwood2** [UoMResearchIT/CapX#20]
Need to add some logging for actions undertaken based on the logged in user for audit purposes.

**Comment by aharwood2** [comment]
Dependent now on #123 being completed so we have different user information available.

**Comment by ianhinder** [comment]
#123 has now been implemented, so this Issue is unblocked.
Is this issue just about internal logging, or also for a user-visible audit log?  e.g. I'd like to be able to see when changes were made to my project, by whom, and what the change was.  Mostly because if something isn't how I expected it to be, it might have been changed for a reason, and I don't want to change it back without talking to the person who made the change.

**Comment by aharwood2** [comment]
I was initial for internal logging but I can easily make it about both. In general, the policy is that only the project owner should be making changes the entries in CapX unless they are off for some reason and a change needs to be made in their absence. If that is the case then it should be being communicated to the owner via email or Teams for them to see when they return. Do you feel this is not happening?

**Comment by ianhinder** [comment]
I had a project where someone was assigned, and I don't remember seeing them as assigned when I took over the project.  I don't know if this is because someone else assigned the person in the mean time, or if I just missed or forgot about the fact that that person was there.  Other than that, I don't think people are going around changing other peoples' projects.

Since projects have an owner in CapX, and we now have accounts, we could apply some level of control over who edits what.    e.g. maybe if it's not your project, you need to answer an "are you sure?" question before saving?  Or there is an "unlock" button you have to click if you want to edit fields in someone else's project.  Actually preventing editing would probably be counterproductive, as then you would have more process to go through when someone is off.

**Comment by aharwood2** [comment]
I see, yes. Somewhere in between your suggestions could be that we allow a small number of superusers access to everything and then the managers are just restricted to editing their own stuff based on ownership? Would you mind @ianhinder creating a new issue for edit restrictions on projects to extend to managers too based on ownership?

**Comment by ianhinder** [comment]
Note that at the moment, everyone can edit the project owner field, and my impression is that I have often found the field to be blank.  Maybe the field should be automatically set to the user who creates the project, and should not be modifiable except by the project owner or a super-user?

**Comment by aharwood2** [comment]
> Note that at the moment, everyone can edit the project owner field, and my impression is that I have often found the field to be blank. Maybe the field should be automatically set to the user who creates the project, and should not be modifiable except by the project owner or a super-user?

It could be setup that way, sure. The reason why most are blank is because the field was only added recently so 100s of projects were already in the DB and so the column value was blank by default when running the migration. I think most people have been adding their names now when they come to update their projects so it should be improving!

**Comment by aharwood2** [comment]
I've added some basic logging that logs broadly what the user is doing i.e. "<username>: Edited RTP-111" or "<username>: Deleted task Development". This will help a little with traceability but I suspect what you really want here Ian is something of an audit log which tell you precisely what field was changed and from what value to what. This would obviously require more work and I think was not really the intention of this issue so perhaps needs to be raised as a separate issue called something like "DB Change Audit Log". If we were using SQL Server then we could just use the server's features to capture and report this but as we are using SQLite, it probably doesn't have something so sophisticated. We can perhaps use the standard entity state tracking built into EF Core? Will look into if we create a new issue for this.
